models/status.go

```diff
@@ -0,0 +1,266 @@
+// Copyright 2017 Gitea. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package models
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"code.gitea.io/git"
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/setting"
+	api "code.gitea.io/sdk/gitea"
+
+	"github.com/go-xorm/xorm"
+)
+
+// CommitStatusState holds the state of a Status
+// It can be "pending", "success", "error", "failure", and "warning"
+type CommitStatusState string
+
+// IsWorseThan returns true if this State is worse than the given State
+func (css CommitStatusState) IsWorseThan(css2 CommitStatusState) bool {
+	switch css {
+	case CommitStatusError:
+		return true
+	case CommitStatusFailure:
+		return css2 != CommitStatusError
+	case CommitStatusWarning:
+		return css2 != CommitStatusError && css2 != CommitStatusFailure
+	case CommitStatusSuccess:
+		return css2 != CommitStatusError && css2 != CommitStatusFailure && css2 != CommitStatusWarning
+	default:
+		return css2 != CommitStatusError && css2 != CommitStatusFailure && css2 != CommitStatusWarning && css2 != CommitStatusSuccess
+	}
+}
+
+const (
+	// CommitStatusPending is for when the Status is Pending
+	CommitStatusPending CommitStatusState = "pending"
+	// CommitStatusSuccess is for when the Status is Success
+	CommitStatusSuccess CommitStatusState = "success"
+	// CommitStatusError is for when the Status is Error
+	CommitStatusError CommitStatusState = "error"
+	// CommitStatusFailure is for when the Status is Failure
+	CommitStatusFailure CommitStatusState = "failure"
+	// CommitStatusWarning is for when the Status is Warning
+	CommitStatusWarning CommitStatusState = "warning"
```

Comment:
Well, got the initial idea when I had intermittant race-conditions in my tests. Re-running the tests made 'em pass, but you should obviously fix the race-condition, hence "warning" 😉 

The _reason_ I want it thought is mainly for when you're rate-limited. I want it to return "warning" since you **should** fix the issue, but it might not be critical to fix it _now_ 🙂